File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.java
class Solution {
public:
    
    int maxPathSum(TreeNode* root) {
        int maxValue=INT_MIN;
         maxPathDown(root,maxValue);
         return maxValue;
    }
    int maxPathDown(TreeNode *root,int &maxValue)
    {
        if(root==NULL) return 0;
        int left=max(0,maxPathDown(root->left,maxValue));
        int right=max(0,maxPathDown(root->right,maxValue));
        maxValue=max(maxValue,left+right+root->val);
        return max(left,right)+root->val;
    }
};

File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.py
class Solution:
    maxInt = -float('inf')
    def maxPathSum(self, root: Optional[TreeNode]) -> int: 
        def helper(root: TreeNode) -> int:
            if not root:    return 0
            l = max(0,helper(root.left))
            r = max(0,helper(root.right))
            self.maxInt = max(self.maxInt,l + r + root.val)
            return max(l,r) + root.val
        if not root:    return 0
        helper(root)
        return self.maxInt
